### PR TITLE
Integration with S3 for Reading and Uploading CSV Files

### LINKED ISSUE #849: refactor the read_dataset(url) function
Problem: Currently the `read_dataset(url)` function in `ai_drive.py` takes in a URL to a csv file, reads that data and stores it in a local file called `data.csv`. However, due to our team pivoting to using more of S3 to store user uploaded data files, we should make this function integrate with the S3 upload accordingly. 

Solution: If we are given a url to a csv dataset, we should be able to:

1. parse the csv file name from the url itself (or use a generic name that the user knows about)
2. get the csv contents and build temporary file (but do not store it locally)
3. send the temporary csv file to s3

Highly recommend testing this endpoint out (either manually or thru unit tests). eg: try to find url to iris dataset and test the refactored function to see if the file gets uploaded to s3

### PR DESCRIPTION
**Pull request by codingwithsurya**
# read_dataset(url) Integration with S3 for Reading and Uploading CSV Files

**What user problem are we solving?**
Currently, the read_dataset(url) function in ai_drive.py reads a CSV file from a given URL and stores it locally as data.csv. However, as the team is shifting towards using S3 for storing user-uploaded data files, we need to modify this function to integrate with S3 for uploading the CSV files directly to S3 instead of saving them locally.

**What solution does this PR provide?**
This PR refactors the read_dataset(url) function in ai_drive.py to handle CSV files as follows:

- Parses the CSV file name from the URL itself or uses a generic name provided by the user.
- Reads the CSV contents and builds a temporary file in memory (not stored locally).
- Uploads the temporary CSV file to an S3 bucket.

The refactored function will now interact with S3 for file storage, which aligns with the team's new approach to data handling.

**Testing Methodology**
I tested the changes in the test_dataset_url_reading function within test_data_processing.py to ensure that the integration with S3 for reading and uploading CSV files worked as expected. The tests included the following:

Tested with jobs.csv URL: Executed the test_dataset_url_reading function with a URL pointing to the jobs.csv dataset. The test confirmed that the function could read the CSV contents from the URL, build a temporary file in memory, and successfully upload it to the designated S3 bucket.

Tested with cars.csv URL: Similarly, I tested the function with a URL pointing to the cars.csv dataset. The test validated that the CSV file was read, processed, and uploaded to S3 without any issues.

Both tests passed, demonstrating that the modifications to the read_dataset(url) function and its integration with S3 were successful and did not break existing functionality.

**Any other considerations**

N/A

### CODE REVIEW

**Review by farisdurrani — APPROVED** [review]

**Comment by farisdurrani** [comment]
lgtm, wait for either Karthik or Daniel's approval

**Comment by farisdurrani** [comment]
All the pyright comments are safe to ignore

**Comment by codingwithsurya** [comment]
> lgtm, wait for either Karthik or Daniel's approval

Thanks. Quick question -- in the github actions, any reason why the tests action isn't running in python package (poetry)? is it because when running the code in a GitHub Actions workflow, you need to provide the AWS credentials explicitly to the workflow so that it can access the S3 bucket? When i run the code locally, AWS credentials are picked up and all tests pass.

**Comment by farisdurrani** [comment]
> > lgtm, wait for either Karthik or Daniel's approval
> 
> Thanks. Quick question -- in the github actions, any reason why the tests action isn't running in python package (poetry)? is it because when running the code in a GitHub Actions workflow, you need to provide the AWS credentials explicitly to the workflow so that it can access the S3 bucket? When i run the code locally, AWS credentials are picked up and all tests pass.

Karthik will be able to give a better answer but to my knowledge, our GitHub Actions are not configured to talk to AWS. We'd have to configure it and include the API key (which is something that should be done)

File: backend/common/dataset.py
Comment (karkir0003) [inline backend/common/dataset.py:34]: explain purpose of this line?

File: backend/common/dataset.py
Comment (karkir0003) [inline backend/common/dataset.py:37]: hardcoding. make constant?

Comment (karkir0003) [reply backend/common/dataset.py:37]: or use the existing one from constants.py?

File: backend/tests/test_data_processing.py
Comment (karkir0003) [inline backend/tests/test_data_processing.py]: I get that this unit test passes. However, when we are testing functionality of aws services, we generally dont want to directly invoke boto3 server calls because then your unit tests have a slight coupling to if boto3 is up and running. 

Try looking into a framework called moto (https://github.com/getmoto/moto) for this purpose. 

@codingwithsurya

Comment (karkir0003) [inline backend/tests/test_data_processing.py]: dont understand why this test is doing same work as read_dataset(url)

your test should be along the lines of 

1. call read_dataset(url)
2. verify that the s3 content coming from read_dataset matches up with contents of the file you read from with provided url

File: backend/common/dataset.py
Comment (karkir0003) [inline backend/common/dataset.py]: might be good to separate the "parse csv(url)" logic from uploading to s3. see if our helpers from s3_client.py can be used to do the put temp file in S3 bucket @codingwithsurya

**Comment by karkir0003** [comment]
@codingwithsurya i domt think we have tests that depend on aws at all (except the test case surya is trying to add in this pr). can we look thru the github actions log to confirm the test step got skipped?

**Comment by karkir0003** [comment]
@farisdurrani @codingwithsurya the unit tests did run in poetry action. the build check actually failes bc the new/modified unit tests didnt pass

https://github.com/DSGT-DLP/Deep-Learning-Playground/actions/runs/5733215885/job/15537550803

**Comment by codingwithsurya** [comment]
> @farisdurrani @codingwithsurya the unit tests did run in poetry action. the build check actually failes bc the new/modified unit tests didnt pass
> 
> https://github.com/DSGT-DLP/Deep-Learning-Playground/actions/runs/5733215885/job/15537550803

Hmm. Are you sure? the unit tests pass locally, and if u look through the github action build check, the exception message is triggered. I think it is  related to the AWS credentials not being found.  When code is run locally, the AWS credentials are usually picked up from your local environment or from a configuration file.

However, when running the code in a GitHub Actions workflow, you need to provide the AWS credentials explicitly to the workflow so that it can access the S3 bucket.

**Comment by karkir0003** [comment]
thats true. but because these two modified unit tests directly interface with aws, you can provide creds via github secrets.

**Comment by codingwithsurya** [comment]
> thats true. but because these two modified unit tests directly interface with aws, you can provide creds via github secrets.

oh I see. and thats been configured? (so workflow can access the S3 bucket)

**Comment by karkir0003** [comment]
hmm. might have to configure that in github actions @codingwithsurya

**Comment by karkir0003** [comment]
see training-container.yaml in .github/workflows for configuring aws credentials in github actions

**Comment by codingwithsurya** [comment]
> hmm. might have to configure that in github actions @codingwithsurya

yea i just did. now every action passes! Do I have permission to merge? @karkir0003

**Comment by karkir0003** [comment]
did you look into other suggestions (ie: the one abt moto) @codingwithsurya

**Comment by codingwithsurya** [comment]
> moto

sorry -- not sure what this means.

**Comment by codingwithsurya** [comment]
> did you look into other suggestions (ie: the one abt moto) @codingwithsurya

just integrated moto and it works. @karkir0003 can u review for merging?

**Comment by codingwithsurya** [comment]
@karkir0003 integrated ur suggestions

**Review by karkir0003 — APPROVED** [review]

**Comment by karkir0003** [comment]
@codingwithsurya lgtm

**Pull request merged**
